### ISSUE TITLE
BUG: hash_pandas_object ignores column name values

### ISSUE THREAD

**Issue by lmeyerov** [pandas-dev/pandas#46705]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import hashlib, pandas as pd

df = pd.DataFrame({'s': ['a', 'b', 'c', 'd'], 'd': ['b', 'c', 'd', 'e'], 'i': [0, 2, 4, 6]})

df_renamed = df.rename(columns={'s': 'ss'})

hash_df = hashlib.sha256(pd.util.hash_pandas_object(df, index=True).values).hexdigest()
hash_df_renamed = hashlib.sha256(pd.util.hash_pandas_object(df_renamed, index=True).values).hexdigest()

assert hash_df != hash_df_renamed
```

### Issue Description

When hashing a df, the column names are ignored

### Expected Behavior

I expected dfs with diff column names to hash differently. If ignoring col names is desired, I'd expect a default-off flag for ignoring col names in the hash calc.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.7.13.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.144+
Version          : #1 SMP Tue Dec 7 09:58:10 PST 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.5
numpy            : 1.21.5
pytz             : 2018.9
dateutil         : 2.8.2
pip              : 21.1.3
setuptools       : 57.4.0
Cython           : 0.29.28
pytest           : 3.6.4
hypothesis       : None
sphinx           : 1.8.6
blosc            : None
feather          : 0.4.1
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 5.5.0
pandas_datareader: 0.9.0
bs4              : 4.6.3
bottleneck       : 1.3.4
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : 0.13.3
pyarrow          : 6.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.4.32
tables           : 3.7.0
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 1.1.0
xlwt             : 1.3.0
numba            : 0.51.2
</details>

**Comment by rhshadrach** [comment]
By doing `.values` you are getting the underlying NumPy array. This doesn't have column names.

**Comment by rhshadrach** [comment]
E.g.

```
df = pd.DataFrame({'s': ['a', 'b', 'c', 'd'], 'd': ['b', 'c', 'd', 'e'], 'i': [0, 2, 4, 6]})
print(pd.util.hash_pandas_object(df, index=True).values)
```

gives

```
[ 7494370580092021340  5889295906979463579  1670643539686186963
 10982910506778653110]
```

**Comment by lmeyerov** [comment]
That makes sense

We're currently doing a second explicit round of hashing to include `df.columns` -- is there a more natural way to get a full hash?

**Comment by rhshadrach** [comment]
Perhaps this is an odd solution, but one thought is to use `to_parquet` with a byte stream.

```
buffer = BytesIO()
df.to_parquet(buffer)
hash_df = hashlib.sha256(buffer.getbuffer()).hexdigest()
```

This will fail if the DataFrame contains complex objects (e.g. lists) as those can't be handled by parquet.

**Comment by lmeyerov** [comment]
Yeah we already have a a `try/except` here (for arrow conversion) :)

Interestingly, parquet path is about ~10% faster too!

**Comment by rhshadrach** [comment]
@Imeyerov - Can you describe your use case for hashing a DataFrame? In general one typically doesn't hash mutable objects (especially for us in e.g dict keys).

Certainly DataFrames should not be hashable (i.e. `hash(df)` fails), but perhaps there is a use case for pandas to provide a hash value for DataFrame, or at least make obtaining such values more straight foward.

**Comment by lmeyerov** [comment]
Our standard practice is to avoid mutations, so less of a concern here ;-)

Our rough scenario is around memoization optimizations where we can't rely on pointer equality, such as someone rerunning a notebook cell and getting a new df but with old values and avoiding large uploads when that happens: https://github.com/graphistry/pygraphistry/blob/4e71239e177d76a9b8bdca64c6995b7727afc5a5/graphistry/PlotterBase.py#L1684

**Comment by rhshadrach** [comment]
@Gabriel-ROBIN: You can use the BytesIO object directly.

```
df = pd.DataFrame({"a": [1, 1, 1], "b": ["x", "y", "z"], "c": [1, 2, 3]})
buffer = BytesIO()
df.to_parquet(buffer)
with open("temp.parquet", mode="wb") as f:
    f.write(buffer.getvalue())
print(pd.read_parquet("temp.parquet"))
```

**Comment by MarcSkovMadsen** [comment]
Hashing of DataFrames are often done in data app frameworks like Streamlit, Panel etc. Its used for caching purposes to speed up the apps.

**Comment by hagenw** [comment]
> Can you describe your use case for hashing a DataFrame?

Besides caching it might also be useful to check if parts of a dataset have changed.
E.g. when storing a dataframe as a csv file you can calculate a md5 sum of the csv file, which will only change if the content of the csv file changes. But if you store the dataframe in a parquet file, it will have a different md5 sum every time you store the file, even if the underlying data is the same (that is the way parquet files are designed). In this case, you could calculate a hash sum based on the actual dataframe and store this inside the parquet file, and can use that to track if the content of the file has changed.

**Comment by matiaslindgren** [comment]
@rhshadrach I have a proposed solution https://github.com/pandas-dev/pandas/pull/68545 that includes the column labels in the hash. It doesn't include any metadata, though. Is that still in scope of this issue?

**Comment by rhshadrach** [comment]
@matiaslindgren - thanks for the interest here. Looking at this again, I'm negative on including the columns in `hash_pandas_object`. It's my perspective that this function is meant as a row hash and not for the purpose that is being asked for: to be able to tell if two pandas objects are identical. Even adding columns to the hash, there are a number of other aspects where non-identical DataFrames would still produce the same hashes, e.g. int64 vs int32 dtypes (with the same values) and whether the Index name is set vs unset.

I think it's worth considering adding the functionality being requested here, but not to this function. I think I'm +0 here and would need buy-in from other maintainers.

cc @pandas-dev/pandas-core

**Comment by jbrockmendel** [comment]
I’d like to have a better idea of the use case for hashing. Is it really just “check if anything has changed”? If so, that would require including everything right down to attrs. But in that case I’m skeptical hashing is the right approach; write-time cache invalidation seems more standard?

**Comment by lmeyerov** [comment]
We don't always get reference equality but still want to check for value equality, eg, to skip re-uploading the same data

**Comment by hagenw** [comment]
Most likely you will need to adapt it to your kind of data.

E.g. we have implemented the following hashing in order to decide if a dataframe included in a dataset has changed with regards to its values, datatypes, or column names: https://github.com/audeering/audformat/blob/main/audformat/core/utils.py#L749-L857
